src/data_source/fetching.rs

```diff
@@ -340,22 +365,66 @@ where
     Types: NodeType,
 {
     storage: S,
+    height: u64,
```

Comment:
Did the thing I mentioned before where we keep the height in memory, this is actually pretty important for efficiency particularly for subscription streams, where we do a simple and elegant but not inherently performant thing: generate the subscription from basically an infinite stream of `Fetch` requests. This could be really bad because it would generate a DB query for every single entry in the stream, even though we know that once we reach the end of the current chain, subsequent entries will not actually be in the database and will instead be received by the event stream. This in-memory height allows these fetches to short circuit before they hit the DB, making this simple design just about as efficient as a more complicated thing with separate logic for blocks before and after the end of the chain.